Add GithubCredentialsProvider for start of GitHub Apps support (#4058)
backstage/backstage: docs/plugins/github-apps.md

> # Using GithubApps for backend authentication

Reviewer: ```suggestion
# Using GitHub Apps for Backend Authentication
```

Reviewer: Just noting that this file doesn't have any yaml at the start and isn't exposed. I'll presume that that's intentional.

Author: It's intentional for now until we can dogfood the API in the scaffolder

Backstage can be configured to use GitHub Apps for backend authentication. This

> come with advantages such as higher rate limits and that Backstage can act as an

Reviewer: ```suggestion
comes with advantages such as higher rate limits and that Backstage can act as an
```

application instead of a user or bot account.

It also provides a much clearer and better authorization model as a opposed to
the OAuth apps and their respective scopes.

## Caveats

> - It's not possible to have multiple backstage GitHub Apps installed in the same

Reviewer: ```suggestion
- It's not possible to have multiple Backstage GitHub Apps installed in the same
```

>   GitHub organization be managed by Backstage. We currently don't check through

Reviewer: ```suggestion
  GitHub organization, to be handled by Backstage. We currently don't check through
```

  all the registered GitHub Apps to see which ones are installed for a

>   particular repository. We just respect global Organization installs right now.

Reviewer: ```suggestion
  particular repository. We only respect global Organization installs right now.
```

> - App permissions is not managed by Backstage. They're created with some simple

Reviewer: ```suggestion
- App permissions are not managed by Backstage. They are created with some simple
```

  default permissions which you are free to change as you need, but you will

>   need to update them in the GitHub web console, not in Backstage right now. The

Reviewer: ```suggestion
  currently need to update them in the GitHub web console, not in Backstage. The
```

>   permissions that are defaulted are `metadata:read` and `contents:read`.

Reviewer: ```suggestion
  default permissions are `metadata:read`, and `contents:read`.
```

> - The created GitHub App is private by default, this is most likely what you

Reviewer: ```suggestion
- The created GitHub App is private by default. This is most likely what you
```

  want for github.com but it's recommended to make your application public for

>   GitHub Enterprise in order to share application across your GHE organizations.

Reviewer: ```suggestion
  GitHub Enterprise in order to share the application across your GHE organizations.
```

A GitHub app created with `backstage-cli create-github-app` will have read
access by default. You have to manually update the GitHub App settings in GitHub

> to grant the app more permissions if needed.

Reviewer: this is basically a repetition of the text above, isn't it?

> ### Using the CLI (public GitHub only)

Reviewer: i think this and the ones below should be level 2 `##` right? otherwise they are under `Caveats`

> You can use the `backstage-cli` to create GitHub App' using a manifest file that

Reviewer: ```suggestion
You can use the `backstage-cli` to create a GitHub App using a manifest file that
```

we provide. This gives us a way to automate some of the work required to create

> a GitHub app.

Reviewer: ```suggestion
a GitHub App.
```

You can read more about the `backstage-cli create-github-app` method

> [here](../cli/commands.md#create-github-app)

Reviewer: ```suggestion
[here](../cli/commands.md#create-github-app).
```

> Once you've gone through the CLI command, it should produce a `yaml` file in the

Reviewer: ```suggestion
Once you've gone through the CLI command, it should produce a yaml file in the
```

> root of the project which you can then use as an `include` in your

Reviewer: ```suggestion
root of the project, which you can then use as an `include` in your
```

`app-config.yaml`. You can go ahead and skip to
[here](#including-in-integrations-config) if you've got to this part.

### GitHub Enterprise

> You have to create the GitHub Application manually using these

Reviewer: ```suggestion
You have to create the GitHub App manually using [these
```

> [instructions](https://docs.github.com/en/free-pro-team@latest/developers/apps/creating-a-github-app)

Reviewer: ```suggestion
instructions](https://docs.github.com/en/free-pro-team@latest/developers/apps/creating-a-github-app),
```

as GitHub Enterprise does not support creation of apps from manifests.

> Once the application is created you have to generate a private key for the

Reviewer: ```suggestion
Once the app is created you have to generate a private key for it
```

> application it in a `yaml` file.

Reviewer: ```suggestion
in the form of a yaml file.
```

The yaml file must include the following information. Please note that the
indentation for the `privateKey` is required.

```yaml
appId: 1
clientId: client id
clientSecret: client secret
webhookSecret: webhook secret
privateKey: |
  -----BEGIN RSA PRIVATE KEY-----
  ...Key content...
  -----END RSA PRIVATE KEY-----
```

### Including in Integrations Config

> Once the credentials are stored in a `yaml` file generated by

Reviewer: ```suggestion
Once the credentials are stored in a yaml file generated by
```

`create-github-app` or manually by following the

> [GitHub Enterprise](#gitHub-enterprise) instructions they can be included in the

Reviewer: ```suggestion
[GitHub Enterprise](#gitHub-enterprise) instructions, they can be included in the
```

> `app-config.yaml` under the integrations section.

Reviewer: ```suggestion
`app-config.yaml` under the `integrations` section.
```

```yaml
integrations:
  github:
    - host: github.com
    apps:

>         - $include: example-backstage-app-credentials.yaml

Reviewer: Maybe we need to clarify around security here since this is a file that holds highly sensitive information. How would this really work in a production scenario? Or for a user that just wants to do some local dev? Will the file be checked into VCS? How will it be shipped in CI/CD? Etc

Author: Let's add a line about this file being sensitive in this doc. 

It's should definitely not be checked into git. The idea would be to add `*-credentials.yaml` to the .gitignore template for `create-app`. How the file is delivered to production is highly dependent on your production environment but there's no problems shipping a yaml file with credentials using Vault or by a ConfigMap hooked up to some secret store.

Author: @Rugvip does the above sound sane or have you got other suggestions?

If the above sounds like a plan then we need to alter the file name for create-github-app because it does not append `-credentials` at this point.

Reviewer 2: We already have `*.local.yaml` ignored, so I think it's better to stick to that? You'll really need to include that in a `app-config.local.yaml` config as well, so it's pretty detached from the production config scenario. In the production case you can name them whatever you want and bring them in with the secret management system of choice.

Reviewer 3: I thought about this, I just thought that you kind of get the feeling that the credentials file is then only supposed to be used on your local machine and maybe is just abusing the fact that we have *.local.yaml already in the `.gitignore`. I'm happy either way to be honest, just thought it was a little misleading.

Reviewer: Mm I kinda get the same feeling

Reviewer 2: Yeah agree, but tbh that's more or less the way it works. Those are your local credentials, and the ones you want to upload to secret management just happens to be the same? It would be valid to create a separate private key for use in the production configuration, iirc GitHub supports that.

Author: I created #4107 that adds `-credentials` to the file name and to gitignore. 
I'm happy to change it to .local if you have a strong preference.
